Add End of Maintainance, End of basic support and Yanked warnings with FastAPI + React UI for Airflow 3.1

sunank200 opened the pull request:
Adds End of Maintenance (EOM), End of Basic Support (EOBS) and Yanked warnings to the Airflow UI with dismissal support.

### Changes
- FastAPI endpoints for version status and dismissal (`/version_check/status`, `/dismiss/eom`, `/dismiss/eobs`)
- React UI component for displaying warnings with dismiss functionality
- Database schema updated with separate `eom_dismissed_until` and `eobs_dismissed_until` columns

### Configuration
New settings under `[astronomer]`:
- `eom_warning_opt_out` / `eobs_warning_opt_out` - disable warnings
- `eom_warning_threshold_days` / `eobs_warning_threshold_days` - days before showing warning (default: 30)
- `eom_dismissal_period_days` / `eobs_dismissal_period_days` - dismissal duration (default: 7)

### Notes
- Requires Airflow 3.0+ for FastAPI, 3.1+ for React UI
- I have Used [edge3 EdgeExecutorPlugin](https://github.com/apache/airflow/blob/main/providers/edge3/src/airflow/providers/edge3/plugins/edge_executor_plugin.py) as reference for plugin structure
- Migration included in existing `c7f8e9a2b3d4` revision (not yet released)

### Testing:
Testing it by building it locally and changing value in DB.
1. End of basic support

<img width="1728" height="1041" alt="Screenshot 2026-01-14 at 8 52 25 PM" src="https://github.com/user-attachments/assets/58b4b3b7-9998-4c15-9a02-0d272beb9245" />
<img width="1727" height="1030" alt="Screenshot 2026-01-14 at 9 17 23 PM" src="https://github.com/user-attachments/assets/863a97fa-20a7-4c85-8bb9-cc8b2997436e" />



2. Yanked
<img width="1728" height="1035" alt="Screenshot 2026-01-14 at 9 17 52 PM" src="https://github.com/user-attachments/assets/c6297aec-87c9-4dd4-90ab-0bde92007284" />




closes: [#2737](https://github.com/astronomer/astro-runtime/issues/2737)

sunank200 committed:
Add FastAPI endpoints for version check status and dismissal

sunank200 committed:
Add React UI for version warnings with dismissal support

sunank200 committed:
Add EOM/EOBS warning support with separate dismissal tracking

sunank200 committed:
Apply pre-commit fixes and update package_data

sunank200 committed:
Fix React icon imports and update routes

sunank200 committed:
Restore missing version_compat.py file after rebase

sunank200 committed:
Fix multiple Alembic heads by restoring placeholder migration and linking Airflow 3 migration

sunank200 committed:
Make Flask import lazy to avoid requiring Flask in Airflow 3

jedcunningham commented on astronomer/airflow/version_check/migrations/versions/b5ad49d1f9b4_add_placeholder_migration.py:
Why did this get added back?

sunank200 replied:
After rebasing with main, I got this error while testing this change with runtime:
```
alembic.util.exc.CommandError: Multiple heads are present; please specify a single target revision
alembic.script.revision.MultipleHeads: Multiple heads are present for given argument 'head'; b5ad49d1f9b4, c7f8e9a2b3d4
```

The placeholder migration `b5ad49d1f9b4` was removed from code but still exists in the database while testing with astro-runtime. The new migration `c7f8e9a2b3d4` had `down_revision = None`, creating two root migrations (multiple heads).

This resolves the multiple heads error.

jedcunningham replied:
Just reset your db. We don't need to carry around an empty migration forever.

sunank200 replied:
I debugged this locally. Locally I had the old migration file is in `build/lib/`. And alembic is scanning the installed package location (from PACKAGE_DIR), which still includes the old migration. That caused this previously.

sunank200 replied:
Changed it now.

jedcunningham commented on astronomer/airflow/version_check/version_api/app.py:
```diff
@@ -0,0 +1,47 @@
+            "checking version status, warnings (End of Maintenance, End of Basic Support, Yanked), "
+            "and dismissing warnings."
+        ),
+        version="1.0.0",
```
We should use the plugin version I think?

sunank200 replied:
Changed it.

jedcunningham commented on astronomer/airflow/version_check/version_api/app.py:
```diff
@@ -0,0 +1,47 @@
+            "/res",
+            StaticFiles(directory=www_res.absolute(), html=True),
+            name="version_check_res_files",
+        )
```
```suggestion
```

yagni

sunank200 replied:
removed it

jedcunningham commented on astronomer/airflow/version_check/www/src/components/DismissButton.tsx:
```diff
@@ -0,0 +1,68 @@
+      onClick={() => dismissMutation.mutate()}
+      loading={dismissMutation.isPending}
+      aria-label="Dismiss warning"
+      title="Dismiss warning for 7 days"
```
Isnt the length configurable though?

sunank200 replied:
Changed it

jedcunningham commented on astronomer/airflow/version_check/www/index.html:
Do we need this?

sunank200 replied:
The index.html file is only used for local development with the vite dev server.

sunank200 replied:
Removed it

jedcunningham commented on astronomer/airflow/version_check/www/src/dev.tsx:
Do we need this?

sunank200 replied:
It was only used with index.html for local development. The production build uses main.tsx as the entry point, so dev.tsx isn't needed.

sunank200 replied:
Removed it

jedcunningham commented on astronomer/airflow/version_check/www/src/theme.ts:
```diff
@@ -0,0 +1,54 @@
+  theme: {
+    tokens: {
+      colors: {
+        brand: {
```
What are these for?

sunank200 replied:
The brand colour tokens aren't used. The theme now only includes the semantic tokens (warning and critical) used by the warning banners. Components use Chakra UI's built-in status colours (warning/critical).

jedcunningham commented on astronomer/airflow/version_check/plugin.py:
```diff
@@ -1,24 +1,86 @@
     name = "astronomer_version_check"
 
+    # Register FastAPI app and React app only on API server and for Airflow 3.0+
+    if RUNNING_ON_APISERVER and AIRFLOW_V_3_0_PLUS:
```
We only want to support 3.1+ anyways, so you can simplify this.

sunank200 replied:
Simplified this

sunank200 committed:
Use plugin version instead of hardcoded version in FastAPI app

sunank200 committed:
Use configurable dismissal period in DismissButton tooltip

sunank200 committed:
Remove unused development files and unused brand colors from theme

sunank200 committed:
Remove version check from plugin registration, only support Airflow 3.1+

sunank200 committed:
Remove unused files: testsSetup.ts and src/res/README.md

jedcunningham reviewed (commented):
Definitely needs review from some ui folks too.

jedcunningham commented on astronomer/airflow/version_check/version_api/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
```
```suggestion
```
We aren't adding the license header anywhere else, so lets toss it?

sunank200 replied:
Removed it

sunank200 committed:
Remove placeholder migration and make c7f8e9a2b3d4 the root migration

pierrejeambrun commented on README.md:
```diff
@@ -30,16 +42,42 @@ variables prefixed with `AIRFLOW__ASTRONOMER__`.
+- `POST /version_check/ui/dismiss/eobs` - Dismiss EOBS warning for the configured period
 
-  Sets the threshold for showing EOL warnings. The default is 30 days.
+API documentation is available at `/version_check/docs` (Swagger UI) and `/version_check/redoc` (ReDoc).
```
There is no step currently in the packaging process to build the UI assets. (bundle, pnpm install + pnpm build).

I guess what we need is to have assets built and available in the final wheel.

(Basically add a `pnpm install; pnpm build` command in there)

sunank200 replied:
Updated the step to build UI assets in README.md

pierrejeambrun commented on astronomer/airflow/version_check/migrations/versions/c7f8e9a2b3d4_create_airflow3_tables.py:
```diff
@@ -48,7 +48,8 @@ def upgrade() -> None:
         sa.Column("end_of_maintenance", UtcDateTime(timezone=True), nullable=True),
         sa.Column("end_of_basic_support", UtcDateTime(timezone=True), nullable=True),
-        sa.Column("eos_dismissed_until", UtcDateTime(timezone=True), nullable=True),
+        sa.Column("eom_dismissed_until", UtcDateTime(timezone=True), nullable=True),
+        sa.Column("eobs_dismissed_until", UtcDateTime(timezone=True), nullable=True),
```
This supposes that nobody has ever run the `c7f8e9a2b3d4` migration on their db. (normally the version plugin check with this migration was never released I guess but I prefer to double check).

(otherwise an extra migration file for those two columns is safer)

sunank200 replied:
Yes it is not released yet.

We discussed it here as well: https://github.com/astronomer/astronomer-airflow-version-check/pull/91#discussion_r2641304114

pierrejeambrun commented on astronomer/airflow/version_check/version_api/app.py:
```diff
@@ -0,0 +1,40 @@
+        version=plugin.__version__,
+    )
+
+    app.include_router(ui_router, prefix="/ui")
```
Everything is already under `/version_check` I don't think we need an extra `/ui` prefix.

pierrejeambrun replied:
Or maybe this is to explain that this is not 'public' API and people should not rely on that? That works too.

sunank200 replied:
The API endpoints are now:

- GET /version_check/status 
- POST /version_check/dismiss/eom 
- POST /version_check/dismiss/eobs

pierrejeambrun commented on astronomer/airflow/version_check/version_api/datamodels.py:
```diff
@@ -0,0 +1,61 @@
+class DismissRequest(BaseModel):
+    """Request to dismiss a warning."""
+
+    pass  # No additional fields needed, the type is in the URL
```
If not needed, you can delete this model. (It's basically an empty payload?)

sunank200 replied:
Deleted this

pierrejeambrun commented on astronomer/airflow/version_check/version_api/routes.py:
```diff
@@ -0,0 +1,220 @@
+    from astronomer.airflow.version_check.plugin import (
+        eobs_dismissal_period_days,
+        eom_dismissal_period_days,
+    )
```
I'm not sure about all those local imports, I would be in favor of moving them to the top of the file.

sunank200 replied:
Changed this

pierrejeambrun commented on astronomer/airflow/version_check/version_api/routes.py:
Compared to the airflow v2 implementation we are missing the `permission check`. At least for the dismiss part, this should probably not be accessible from unautheticated user.

pierrejeambrun replied:
And all endpoint should required an authenticated user I guess?

sunank200 replied:
Used `GetUserDep` from `airflow.api_fastapi.core_api.security`.
Updated all three route handlers to use `_user: GetUserDep` as a dependency parameter.

pierrejeambrun replied:
In the 2.x version we had something like `@has_access([("can_dismiss", 'UpdateAvailable')])`, as if there was a specific permission `can_dismiss`.

In the 3.x implementation now anybody authenticated will be able to dismiss are we good with that?

sunank200 replied:
Added requires_access_configuration(method="PUT") to both /dismiss/eom and /dismiss/eobs endpoints

pierrejeambrun commented on astronomer/airflow/version_check/www/src/main.tsx:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Main plugin component
+ */
+const PluginComponent: FC<PluginComponentProps> = () => {
```
Normally you don't need explicit typing like this everywhere for FC.
```suggestion
const PluginComponent = () => {
```

sunank200 replied:
Removed explicit FC type annotations from React components

pierrejeambrun commented on astronomer/airflow/version_check/www/src/theme.ts:
```diff
@@ -0,0 +1,38 @@
+          text: { value: { base: "#991b1b", _dark: "#fee2e2" } },
+        },
+      },
+    },
```
Why did we replace the whole generated theme, with a few specific values like this?

Is that to match the astronomer branding? Otherwise keeping the default file generated by the boilerplate code is better

sunank200 replied:
I wanted to add warning types actually.

sunank200 replied:
Checked now. Default is also similar. So removed this

pierrejeambrun commented on astronomer/airflow/version_check/plugin.py:
```diff
@@ -11,14 +14,66 @@
+            ]
+        except Exception:
+            pass
+
```
I don't think we need to do that. This version check plugin will only be installed on airflow v3+ instances. (I don't have the full picture but to me only new runtime image AF3.1+ will have the version check plugin installed)

Jens did that on edge provider because edge provider needed to be backward compatible.

sunank200 replied:
I removed the Airflow 3.1 version check in a previous commit. I was just trying to see if the API server is running. However, I suppose it is not required in this case.

sunank200 replied:
Changed it.

pierrejeambrun commented on astronomer/airflow/version_check/update_checks.py:
```diff
@@ -440,13 +486,13 @@ def available_update(self):
-    def available_eol(self) -> dict[str, Any] | None:
-        """Check if there is an EOL notice for the current version of Astronomer Runtime."""
+    def available_eom(self) -> dict[str, Any] | None:
+        """Check if there is an End of Maintenance (EOM) notice for the current version."""
```
Do we need to expose a `UpdateAvailableHelper` ? Everything is self contained and accessible through the added API endpoints. Having an external helper might not be necessary anymore?

sunank200 replied:
Yes it it not necessary now. Removed it

pierrejeambrun commented on astronomer/airflow/version_check/www/package.json:
```diff
@@ -0,0 +1,75 @@
+    "build": "vite build",
+    "build:types": "tsc --p tsconfig.lib.json",
+    "build:lib": "vite build",
+    "lint": "eslint --quiet && tsc --p tsconfig.app.json",
```
We might want to keep the generated `eslint` file and add the linting to the CI, so we have a bare minimum of code quality for the UI plugin and make sure it doesn't break easily.

sunank200 replied:
I have create eslint.config.js with TypeScript, React, and Prettier support along with pre-commit

pierrejeambrun reviewed (commented):
A few suggestions, but it's working as expected and looking good.

Thanks a lot for that big effort.

sunank200 committed:
Remove redundant /ui prefix from version_check API endpoints

sunank200 committed:
Remove unused DismissRequest model

sunank200 committed:
Move plugin imports to top of routes.py

sunank200 committed:
Remove explicit FC type annotations from React components

sunank200 committed:
Since this plugin is only for Airflow 3.1+, we don't need the
conditional check or try/except block. Register FastAPI and React
apps directly as class attributes.

sunank200 committed:
Add ESLint configuration and CI linting for UI plugin

sunank200 committed:
Install Node.js and pnpm before running prek in CircleCI.

sunank200 committed:
Add authentication checks to all API endpoints

sunank200 committed:
Remove UpdateAvailableHelper class

sunank200 committed:
Remove liscense

sunank200 committed:
Add explanation of warning threshold vs dismissal period in README

sunank200 committed:
Add building and packaging instructions for UI assets

jedcunningham commented on .circleci/config.yml:
```diff
@@ -47,6 +47,20 @@ jobs:
+      - run:
+          name: Install Node.js and pnpm for pre-commit hooks
+          command: |
+            curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.0/install.sh | bash
```
Why are we using a version from 2021?

jedcunningham replied:
Related, can we get away with using the `-node` variant of the python image?

sunank200 replied:
Changed it to use `-node` variant

jedcunningham commented on .circleci/config.yml:
```diff
@@ -47,6 +47,20 @@ jobs:
+            curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.0/install.sh | bash
+            export NVM_DIR="$HOME/.nvm"
+            [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
+            nvm install 22
```
Should we use 24, the latest lts?

sunank200 replied:
Now using `-node` variant

jedcunningham commented on astronomer/airflow/version_check/version_api/routes.py:
```diff
@@ -0,0 +1,222 @@
+    WarningType,
+)
+
+ui_router = AirflowRouter(tags=["UI"])
```
Is this the right tag?

We don't need to use AirflowRouter (probably isn't right to?)?

jedcunningham replied:
Do we even need a router? Might make it easier to have split routes/app?

sunank200 replied:
I had used AirflowRouter as it was also used in edge3 here: https://github.com/apache/airflow/blob/main/providers/edge3/src/airflow/providers/edge3/worker_api/routes/ui.py#L54

sunank200 replied:
AirflowRouter is a subclass of APIRouter that adds a convenience method api_route for setting operation_id automatically as per https://github.com/apache/airflow/blob/main/airflow-core/src/airflow/api_fastapi/common/router.py#L27

pierrejeambrun replied:
Yeah it's better to have that, so the swagger is pretty.

sunank200 replied:
So, I am keeping AirflowRouter then

pierrejeambrun replied:
Just discussed with Jed, it's better to remove it and use the default router

sunank200 replied:
Sure. I will do the change now.

sunank200 replied:
I have now removed the router and define routes directly in app

sunank200 replied:
Please review

jedcunningham commented on astronomer/airflow/version_check/version_api/routes.py:
```diff
@@ -0,0 +1,222 @@
+            # Check if dismissed
+            if current_version.eobs_dismissed_until and now < current_version.eobs_dismissed_until:
+                pass  # Dismissed, skip to EOM check
+            else:
```
Just invert it so we fall out of the EOBS check vs having to us pass.

sunank200 replied:
Changed it

jedcunningham commented on astronomer/airflow/version_check/update_checks.py:
```diff
@@ -34,6 +32,13 @@
 
 
 def has_access_(permissions: Sequence[tuple[str, str]]) -> Callable[[T], T]:
```
We don't need to support af2, so we can drop all this.

sunank200 replied:
removed it

jedcunningham commented on astronomer/airflow/version_check/version_compat.py:
No longer need this file.

sunank200 replied:
removed it

jedcunningham commented on .pre-commit-config.yaml:
```diff
@@ -32,3 +32,12 @@ repos:
+        language: system
+        types: [file]
+        files: ^astronomer/airflow/version_check/www/.*\.(ts|tsx|js|jsx)$
+        pass_filenames: false
```
We don't want to pass the filenames?

sunank200 replied:
Changed the approach

pierrejeambrun commented on astronomer/airflow/version_check/www/src/theme.ts:
```diff
@@ -0,0 +1,3 @@
+import { createSystem, defaultConfig } from "@chakra-ui/react";
+
+export const system = createSystem(defaultConfig);
```
This is why your plugin doesn't fit well with 'core' theme in dark mode and other edge cases. This will default to chakra base theme, instead of using the Core theme. You need to keep the generated file unchanged. (which is basically a full copy of core theme), and I think this will make things look better.

sunank200 replied:
Fixed. Copied the full Airflow core theme from airflow-core/src/airflow/api_fastapi/auth/managers/simple/ui/src/theme.ts to match the pattern used by other plugins. The plugin now uses the complete theme configuration for proper dark mode integration.

sunank200 replied:
<img width="1726" height="1045" alt="Screenshot 2026-01-14 at 9 16 21 PM" src="https://github.com/user-attachments/assets/eefacd3b-f6ca-46c4-b153-a982dc628e34" />

Changed now

pierrejeambrun reviewed (commented):
Beside Jed comments.

LGTM, just one comment for fixing the theme

sunank200 committed:
Use -node variant in CircleCI and invert EOBS dismissal condition

sunank200 committed:
Remove pass_filenames: false from lint-ui hook

sunank200 committed:
Remove Airflow 2 compatibility code

sunank200 committed:
Add pre-commit hooks for UI linting and asset compilation

sunank200 committed:
Fix compile-assets hook: exclude lock files from hash calculation

sunank200 committed:
Add permission checks for dismiss endpoints

sunank200 committed:
Add MANIFEST.in and pnpm-lock.yaml

sunank200 committed:
Add .pnpm-store to .gitignore

sunank200 committed:
Use full Airflow core theme for proper dark mode integration

pierrejeambrun reviewed (approved):
API and UI wise this looks correct to me.

Might be great to have another review from Jed since this one is pretty big, it's easy to miss something.

jedcunningham commented on .circleci/config.yml:
```diff
@@ -24,7 +24,7 @@ workflows:
   build:
     docker:
-      - image: cimg/python:3.9
+      - image: cimg/python:3.9-node
```
3.9 is EOL.

sunank200 replied:
Changed it to  `cimg/python:3.11-node`

jedcunningham commented on .circleci/config.yml:
```diff
@@ -47,6 +47,14 @@ jobs:
+      - run:
+          name: Install Node.js and pnpm for pre-commit hooks
+          command: |
+            # Check Node.js version that comes with the image
```
```suggestion
          name: Install pnpm and project dependencies
          command: |
```

sunank200 replied:
Changed it

sunank200 committed:
Remove router and define routes directly in app

jedcunningham commented on scripts/ci/prek/compile_version_check_assets.py:
```diff
@@ -0,0 +1,79 @@
+        old_hash = HASH_FILE.read_text().strip() if HASH_FILE.exists() else ""
+        new_hash = get_directory_hash(WWW_DIR, skip_path_regexps=SKIP_PATH_REGEXPS)
+        if new_hash == old_hash:
+            print(f"The '{WWW_DIR}' directory has not changed! Skip regeneration.")
```
Doesn't pre-commit basically give us this out of the box? And `--all-files` on ci gives us safety?

sunank200 replied:
Removed the hash checking logic. Pre-commit's file tracking and CI's --all-files flag are sufficient.

jedcunningham commented on scripts/ci/prek/compile_version_check_assets.py:
```diff
@@ -0,0 +1,79 @@
+        print("Old hash: " + old_hash)
+        print("New hash: " + new_hash)
+    else:
+        shutil.rmtree(DIST_DIR, ignore_errors=True)
```
Why are we removing it if it doesn't exist?

sunank200 replied:
Removed this logic

jedcunningham commented on scripts/ci/prek/compile_version_check_assets.py:
```diff
@@ -0,0 +1,79 @@
+
+    print("### Installing pnpm dependencies ###")
+    subprocess.check_call(
+        ["pnpm", "install", "--frozen-lockfile", "--config.confirmModulesPurge=false"],
```
Do we need the `pnpm install` in the CI job then?

sunank200 replied:
Removed the duplicate pnpm install from CI. The pre-commit hook script already handles it.

jedcunningham commented on .circleci/config.yml:
```diff
@@ -56,6 +64,11 @@ jobs:
           paths:
             - ~/.cache/prek
+      - run:
+          name: lint_ui
```
Is this not covered with the `lint-ui` prek hook? If not, feels like it should be and not be a separate thing.

jedcunningham replied:
It should at least be its own prek hook.

sunank200 replied:
That's right. Removed the redundant `lint_ui` step. The prek run --all-files command already runs the lint-ui pre-commit hook, which performs the same checks.

jedcunningham commented on .circleci/config.yml:
```diff
@@ -47,6 +47,14 @@ jobs:
+          command: |
+            # Check Node.js version that comes with the image
+            node --version
+            sudo npm install -g pnpm
```
If prek is managing the env for the compile/lint hooks, do we even need pnpm globally here?

sunank200 replied:
That's right. I have removed global pnpm installation. prek manages it via additional dependencies.

jedcunningham commented on MANIFEST.in:
Does uv_build use MANIFEST.in?

jedcunningham replied:
Ugh, in flight change: https://github.com/astronomer/astronomer-airflow-version-check/pull/92

sunank200 replied:
Yes, uv_build uses MANIFEST.in, so consolidating everything there works for both setuptools and uv_build.

jedcunningham replied:
Are you sure uv_build uses MANIFEST.in? Isn't that just a setuptools thing? Have you actually tried it?

sunank200 replied:
I checked. uv build uses setuptools (per pyproject.toml), and setuptools respects MANIFEST.in, so the files are included. Here build requires ["setuptools", "wheel"], so uv build uses setuptools.
When I ran uv build, the output showed setuptools commands (running bdist_wheel, copying), confirming setuptools is used.
I also checked the wheel:  www/dist showed main.umd.cjs and main.d.ts included. It matches MANIFEST.in's recursive-include astronomer/airflow/version_check/www/dist *.

jedcunningham commented on www-hash.txt:
Semi-related to the hash/prek commients, but it feels like this shouldn't be checked in so CI always recompiles.

sunank200 replied:
Removed it and added it to .gitignore

jedcunningham commented on setup.py:
```diff
@@ -76,6 +76,10 @@ def find_version(*paths):
         "": ["LICENSE"],
         "astronomer.airflow.version_check.templates": ["*"],
         "astronomer.airflow.version_check.static": ["*"],
+        "astronomer.airflow.version_check": ["alembic.ini"],
```
Let's reduce duplication here and in MANIFEST.in - that'll make it easier to reconcile with the move to uv_build.

sunank200 replied:
Done. Removed package_data from setup.py and consolidated everything in MANIFEST.in. This works with both setuptools and uv_build.

sunank200 committed:
Update Python version and simplify asset compilation

sunank200 committed:
Remove redundant CI steps for UI linting and pnpm installation

sunank200 committed:
Remove duplication between setup.py and MANIFEST.in

sunank200 committed:
Fix static file serving and packaging
